RecoEgamma/EgammaTools/src/HGCalShowerShapeHelper.cc

```diff
@@ -0,0 +1,284 @@
+                                           double minHitET,
+                                           int minLayer,
+                                           int maxLayer,
+                                           DetId::Detector subDet) {
+  // Safety checks
+  nLayer_ = maxLayer - minLayer + 1;
+  assert(nLayer_ > 0);
+
+  minHitE_ = minHitE;
+  minHitET_ = minHitET;
+  minHitET2_ = minHitET * minHitET;
+  minLayer_ = minLayer;
+  maxLayer_ = maxLayer;
+  subDet_ = subDet;
+
+  setFilteredHitsAndFractions(hitsAndFracs);
+
+  setLayerWiseStuff();
+}
+
+void HGCalShowerShapeHelper::setPFRecHitPtrMap(const std::vector<reco::PFRecHit> &recHits) {
+  pfRecHitPtrMap_.clear();
+
+  for (const auto &recHit : recHits) {
+    pfRecHitPtrMap_[recHit.detId()] = &recHit;
+  }
+}
+
+void HGCalShowerShapeHelper::setFilteredHitsAndFractions(const std::vector<std::pair<DetId, float> > &hitsAndFracs) {
+  hitsAndFracs_.clear();
+  hitEnergies_.clear();
+  hitEnergiesWithFracs_.clear();
+
+  for (const auto &hnf : hitsAndFracs) {
+    DetId hitId = hnf.first;
+    float hitEfrac = hnf.second;
+
+    int hitLayer = recHitTools_.getLayer(hitId);
+
+    if (hitLayer > nLayer_) {
+      continue;
+    }
+
+    if (hitId.det() != subDet_) {
+      continue;
+    }
+
+    if (pfRecHitPtrMap_.find(hitId.rawId()) == pfRecHitPtrMap_.end()) {
+      continue;
+    }
+
+    reco::PFRecHit recHit = *pfRecHitPtrMap_[hitId.rawId()];
+
+    if (recHit.energy() < minHitE_) {
+      continue;
+    }
+
+    if (recHit.pt2() < minHitET2_) {
+      continue;
+    }
+
+    // Fill the vectors
+    hitsAndFracs_.push_back(hnf);
+    hitEnergies_.push_back(recHit.energy());
+    hitEnergiesWithFracs_.push_back(recHit.energy() * hitEfrac);
+  }
+}
+
+void HGCalShowerShapeHelper::setLayerWiseStuff() {
+  layerEnergies_.clear();
+  layerEnergies_.resize(nLayer_);
+
+  layerCentroids_.clear();
+  layerCentroids_.resize(nLayer_);
+
+  centroid_.SetXYZ(0, 0, 0);
+
+  int iHit = -1;
+  double totalW = 0.0;
+
+  // Compute the centroid per layer
+  for (const auto &hnf : hitsAndFracs_) {
+    iHit++;
+
+    DetId hitId = hnf.first;
+
+    double weight = hitEnergies_[iHit];
+    totalW += weight;
+
+    const auto &hitPos = recHitTools_.getPosition(hitId);
+    ROOT::Math::XYZVector hitXYZ(hitPos.x(), hitPos.y(), hitPos.z());
+
+    centroid_ += weight * hitXYZ;
+
+    int hitLayer = recHitTools_.getLayer(hitId) - 1;
+
+    layerEnergies_[hitLayer] += weight;
+    layerCentroids_[hitLayer] += weight * hitXYZ;
+  }
+
+  int iLayer = -1;
+
+  for (auto &centroid : layerCentroids_) {
+    iLayer++;
+
+    if (layerEnergies_[iLayer]) {
+      centroid /= layerEnergies_[iLayer];
+    }
+  }
+
+  if (totalW) {
+    centroid_ /= totalW;
+  }
+}
+
+const double HGCalShowerShapeHelper::getCellSize(DetId detId) {
+  double siThickness = recHitTools_.getSiThickness(detId);
+
+  return siThickness < 150 ? kHDWaferCellSize_ : kLDWaferCellSize_;
```

Comment:
can `150` be taken from somewhere else in HGCAL ?
even better, is there a way to get the cell size directly from HCAL ?

Reply:
@rovere ?

Reply:
> Implemented all the review comments except [#32519 (comment)](https://github.com/cms-sw/cmssw/pull/32519#discussion_r545015773).
> @rovere It was not clear to me if/how these numbers can be read from `getNewCorners`.

Ciao @SohamBhattacharya I do not recall the exact meaning of the numbers returned.
Maybe @bsunanda could tell us a bit more. Or, even better, maybe he knows if there's a way to ask a silicon cell in HGCAL about its side or apothem? I know we do have the full length of the wafer and the number of cells along that side, but I'm not sure we do expose directly the dimension of the cell.